doc/developer/design/20240205_cluster_specific_optimization.md

```diff
@@ -0,0 +1,283 @@
+can already be configured in a layered way:
+
+1. A top layer of settings bound from `SystemVars`.
+2. A layer of `EXPLAIN`-specific overrides bound from the `ExplainContext`,
+   which is available when the `Coordinator` methods that drive the optimization
+   process are initiated from `sequence_explain_plan`.
+
+Since all `Optimizer::new(...)` constructor calls in the `Coordinator` happen at
+a time where the target cluster for the optimized statement is already resolved,
+we can just add a new layer for cluster-specific overrides between (1) and (2).
+
+## Minimal Viable Prototype
+
+<!--
+Build and share the minimal viable version of your project to validate the
+design, value, and user experience. Depending on the project, your prototype
+might look like:
+
+- A Figma wireframe, or fuller prototype
+- SQL syntax that isn't actually attached to anything on the backend
+- A hacky but working live demo of a solution running on your laptop or in a
+  staging environment
+
+The best prototypes will be validated by Materialize team members as well
+as prospects and customers. If you want help getting your prototype in front
+of external folks, reach out to the Product team in #product.
+
+This step is crucial for de-risking the design as early as possible and a
+prototype is required in most cases. In _some_ cases it can be beneficial to
+get eyes on the initial proposal without a prototype. If you think that
+there is a good reason for skpiping or delaying the prototype, please
+explicitly mention it in this section and provide details on why you you'd
+like to skip or delay it.
+-->
+
+Nothing is done yet, but once we agree on the SQL extensions the changes to get
+something working end-to-end should be done quite quickly.
+
+## Alternatives
+
+<!--
+What other solutions were considered, and why weren't they chosen?
+
+This is your chance to demonstrate that you've fully discovered the problem.
+Alternative solutions can come from many places, like: you or your Materialize
+team members, our customers, our prospects, academic research, prior art, or
+competitive research. One of our company values is to "do the reading" and
+to "write things down." This is your opportunity to demonstrate both!
+-->
+
+### Managing cluster-specific parameters with LaunchDarkly
+
+Instead of extending the `CREATE CLUSTER` syntax we can also figure out a way to
+enable feature flags on a per-cluster basis through LaunchDarkly.
+
+The approach is rejected (for now) because it requires more substantial changes
+to the current state of our LaunchDarkly setup in `main`. Basically at the
+moment we pull parameters from LaunchDarkly in a loop using a fixed
+`ld::Context`[^1] that consists of:
+
+1. A context of `kind = environment` that models the current Materialize
+   environment.
+2. A context of `kind = organization` that models the environment owner.
+
+If we want to provide cluster-specific configuration through LaunchDarkly, we
```

Comment:
I just want to confirm why this is the hard part: we set up the `system_parameter_sync` loop when environmentd starts up, but now we'd have to notice when clusters come and go and alter the loop (which only has a handle on `Client`)... right?

Reply:
Yes. We don't have to alter / rebuild the connection when the set of clusters changes, but we'll need to build a different LaunchDarkly context type (something which is part of their API) for each cluster make super that the per-cluster parameters are somehow reflected in `SystemParameterFrontend::pull` and `SystemParameterBackend::push`.

Reply:
Agreed that trying to use LD for this seems somewhat challenging. It seems we'd be going off the beaten path with LD if we created a context per cluster, and had each one of those contexts sync with LD separately. Like, having 25 clusters is totally reasonable, but having 25 sync loops with LD is definitely something that would raise eyebrows on their side if we were to tell our SA about it.

That said, having typed it out ... it's not _that_ crazy, and has some appealing properties...

Regardless, I think we ultimately might want to allow users to control these feature flags (within reason! only the ones that we think are safe to control), and so I think the `FEATURES` syntax will not be wasted work even if we one day decide it's worth rigging up LD to support per-cluster features.

Reply:
> having 25 sync loops with LD is definitely something that would raise eyebrows on their side if we were to tell our SA about it.

The sync loop does not translate to explicit requests to the LaunchDarkly servers because we run a server SDK, so all updates for all contexts are always pushed to every client. It means that we'll need to some CPU cycles and memory per cluster (approximately `sizeof(SystemVars) * |clusters|`). but this doesn't sound too scary.

What is not clear to me is how best to save the cluster-specific `SystemVars` copies in the catalog and what API the `SystemParameterBackend` will use to save those changes.

Reply:
Ah! That's actually quite intriguing. There is prior art for this in PostgreSQL: you can set (most) system parameters at the database level, and they take precedence in order of system < database < role < session, or something lke that. I'm confident we could work cluster-specific parameters into the `SystemVars` framework with some elbow grease.

That said, targeting via LD has a bunch of thorny questions we'd need to consider. Do we target by cluster IDs or  names? I like the sound of IDs a bit better than names, but then it's difficult to preinstall LD rules for a new cluster you're about to create. It also seems quite a bit less ergonomic to have to use LD _and_ an `mz_system` shell to run these experiments, instead of being able to do it all via an `mz_system` shell.

Reply:
> It also seems quite a bit less ergonomic to have to use LD _and_ an `mz_system` shell to run these experiments, instead of being able to do it all via an `mz_system` shell.

This is the killer argument for me—at the moment the only use case for cluster-specific parameters is for internal experimentation, and for that you need an `mz_system` shell in order to create an unbilled replica either way.

Reply:
I summarized the discussion above in the corresponding "Alternatives" section in the actual design doc.

Reply:
Slightly ooc, but I wonder how the Platform v2 design will influence our LD syncing. If every cluster gets its own `envd`, or there is no `envd` and adapter runs as part of `clusterd`, that means each cluster will open its own connection to LD anyway and know about its cluster context. Seems like we might get per-cluster LD flags for free with pv2?